matrix.py

```diff
@@ -0,0 +1,156 @@
+"""Matrix exercise, Daniel Prince - unity training."""
+
+
+class Matrix:
+    """Perform all sorts of matrix operations."""
+
+    def __init__(self, data):
+        self.data = data
+        # Check if n x n
+        for tup in self.data:
+            if len(self.data) != len(tup):
+                raise ValueError("not a valid matrix. (not nXn)")
+
+    def __str__(self):
+        return str(self.data)
+
+    def __repr__(self):
+        return '{self.__class__.__name__}({self.data})'.format(self=self)
+
+    def __iter__(self):
+        return iter(self.data)
+
+    def __hash__(self):
+        return hash(self.data)
+
+    def __eq__(self, other):
+        return self.data == other.data
+
+    def __ne__(self, other):
+        return self.data != other.data
+
+    def __add__(self, other):
+        """Adds 2 matrices."""
+
+        if isinstance(other, Matrix):
+            if len(self.data) == len(other.data):
+                post_add_list = [self.data[row_i][column_i]
+                                 + other.data[row_i][column_i]
+                                 for row_i in range(len(self.data))
+                                 for column_i in range(len(other.data))]
+                return Matrix(tuple(splitting(post_add_list, len(self.data))))
+
+            else:
+                print("matrices must be equal length")
+
+        else:
+            print("cannot add a scalar")
+
+    def __sub__(self, other):
+        """Subtracts 2 matrices."""
+
+        if isinstance(other, Matrix):
+            if len(self.data) == len(other.data):
+                post_sub_list = [self.data[row_i][column_i]
+                                 - other.data[row_i][column_i]
+                                 for row_i in range(len(self.data))
+                                 for column_i in range(len(other.data))]
+                return Matrix(tuple(splitting(post_sub_list, len(self.data))))
+
+            else:
+                print("matrices must be equal length")
+
+        else:
+            print("cannot subtract from a / a scalar")
+
+    def __mul__(self, other):
+        """Multiplies matrix by scalar/matrix."""
+
+        if isinstance(other, Matrix):
+            if len(self.data) == len(other.data):
+                # Converting the second matrix to columns
+                column_matrix = [tup[index] for index in range(len(self.data))
+                                 for tup in other.data]
+                split_column_matrix = splitting(column_matrix, len(self.data))
+                post_mul = []
+                for tup in self.data:
+                    for split_list in split_column_matrix:
+                        for num, index in \
+                                zip(split_list, range(len(self.data))):
+                            post_mul.append(num * tup[index])
```

Comment:
I believe that what you intended to do here, is given two vectors lets multiply it element by element.
First you could iterate over `zip(split_list, tup)` instead over iterating over a range and accessing it using indexing.
Second you could replace this part by a one line list comprehension.
Finally you might as well sum it, it will save you work later splitting it s=to smaller lists in order to sum these exact values.

I believe after that you would be able to change the rest of the method so it would look more pythonic.